hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java

```diff
@@ -204,6 +205,53 @@ public void createIndexesFromRecordedDefinitions(StoreAdapter partitionStoreAdap
         return indexes;
     }
 
+    /**
+     * Returns known and populated indexes for the given partition. If no
+     * indexes are known or populated, empty array is returned.
+     * <p>
+     * Must be called on partition threads only.
+     */
+    public InternalIndex[] getPopulatedIndexes(int partitionId) {
+        ThreadUtil.assertRunningOnPartitionThread();
+
+        InternalIndex[] indexes = getIndexes();
+
+        int populatedCount = 0;
+        for (InternalIndex index : indexes) {
+            if (index.hasPartitionIndexed(partitionId)) {
+                ++populatedCount;
+            }
+        }
+
+        if (populatedCount == indexes.length) {
+            return indexes;
+        }
+        if (populatedCount == 0) {
+            return EMPTY_INDEXES;
+        }
+
+        // The code below handles a rare case which may happen during indexes
+        // creation/population. E.g., see MapProxySupport.initializeIndexes:
+        // indexes are added one by one, for every index AddIndexOperation is
+        // issued on all partitions. Individual add index operations may
+        // interleave in any order with any operations including add index
+        // operations issued by other members. So for a global index only some
+        // of its associated partitions might be indexed, but that should not
+        // happen for partitioned HD indexes: all known indexes must be either
+        // populated or not populated for a partition in question.
+
+        assert isGlobal();
+        InternalIndex[] populated = new InternalIndex[populatedCount];
+        int counter = 0;
+        for (InternalIndex index : indexes) {
+            if (index.hasPartitionIndexed(partitionId)) {
+                populated[counter++] = index;
+            }
+        }
+        assert counter == populatedCount;
```

Comment:
`counter == populatedCount` doesn't necessarily need to be true as some partitions might have been indexed since we summed the `populatedCount`, right?

Reply:
`populatedCount` should not change since we are summing populated indexes for the given `partitionId` and we should be running on its partition thread, so no other operations can interleave in between.